source/1_engine/rex_renderer_core/src/renderer.cpp

```diff
@@ -1,25 +1,660 @@
+
+                if (!process_render_command(cmd))
+                {
+                    REX_ERROR(LogRendererCore, "Unable to add/exec create constant buffer command");
+                    return ResourceSlot::make_invalid();
+                }
+
+                return resource_slot;
+            }
+
+            //-------------------------------------------------------------------------
+            bool flush(RingBuffer<RenderCommand*>* cmdList)
+            {
+                // Cache executed commands to destroy after we flushed
+                rsl::vector<RenderCommand*> executed_commands;
+
+                RenderCommand** cmd = cmdList->get();
+
+                while (cmd)
+                {
+                    if ((*cmd)->execute() == false)
+                    {
+                        REX_ERROR(LogRendererCore, "Failed to flush commands");
+                        return false;
+                    }
+
+                    executed_commands.push_back(*cmd);
+
+                    cmd = g_ctx.cmd_list.get();
+                }
+
+                // Manually call the destructor of each command
+                // We will reset the stack allocator back to the beginning and override the existing memory
+                // This will not call any dtors while this happens
+                for (RenderCommand* executed_cmd : executed_commands)
+                {
+                    executed_cmd->~RenderCommand();
+                }
+
+                return true;
+            }
+        }
+
+        //-------------------------------------------------------------------------
+        bool initialize(const OutputWindowUserData& userData, s32 maxCommands, s32 maxFrameResources)
+        {
+            UNUSED_PARAM(userData);
+
+            g_ctx.slot_resources.initialize(32);
+            g_ctx.cmd_list.initialize(maxCommands);
+
+            globals::g_default_targets_info.front_buffer_color = g_ctx.slot_resources.next_slot();
+            globals::g_default_targets_info.back_buffer_color = g_ctx.slot_resources.next_slot();
+            globals::g_default_targets_info.depth_buffer = g_ctx.slot_resources.next_slot();
+
+            return backend::initialize(userData
+                , maxFrameResources
+                , globals::g_default_targets_info.front_buffer_color
+                , globals::g_default_targets_info.back_buffer_color
+                , globals::g_default_targets_info.depth_buffer);
+        }
+        
+        //-------------------------------------------------------------------------
+        void shutdown()
+        {
+            globals::g_default_targets_info.back_buffer_color.release();
+            globals::g_default_targets_info.front_buffer_color.release();
+            globals::g_default_targets_info.depth_buffer.release();
+
+            flush();
```

Comment:
Actually this flush should be at the top of this function, we want to clear all the commands that are still in the queue to not leave the graphics device in an invalid state when the application is closed. Once a command is queue it cannot be dequeued so it could happen that the application is already closed and the GPU tries to execute something which is not available anymore => crash

Reply:
linked to https://github.com/New-Creta/Rex/issues/260